Issue 816: Auto-detect CSV Dates

hawkfish opened the pull request:
As discussed in Issue 816, add code to try to detect some common numeric date formats when sniffing a CSV file.

hawkfish committed:
Merge pull request #14 from cwida/master

Merge as of 28 August 2020

hawkfish committed:
Merge pull request #15 from cwida/master

Pull master for CSV dates work

Commit:
Issue 816: DATE format inference

Test the most common numeric date formats
when sniffing the CSV format
if there is no user-specified format.

Commit:
Issue 108: DATE format inference

Test the most common numeric date formats
when there is no user-specified format.

Commit:
Issue 108: DATE format inference

Choose among common date formats
when sniffing CSV format.

Commit:
Merge branch 'hawkfish-csv-dates' of https://github.com/hawkfish/duckdb into hawkfish-csv-dates

Commit:
Issue 816: Table-driven format inference

Rewrite the date format inference code
to be table driven by type. For now
this will only work for temporal types,
but with appropriate interfaces,
it could be generalised to
arbitrary types.

hawkfish committed:
Merge pull request #16 from cwida/master

Issue 816: Pull from master

Commit:
Issue 816: Reformat

Pre-merge reformat.

Commit:
Issue 816: DATE auto csv

Merge branch 'master' into hawkfish-csv-dates

Mytherin commented on src/include/duckdb/execution/operator/persistent/buffered_csv_reader.hpp:
```diff
@@ -84,13 +85,9 @@ struct BufferedCSVReaderOptions {
-	StrpTimeFormat timestamp_format;
-	//! Whether or not a timestamp format is specified
-	bool has_timestamp_format = false;
+	std::map<LogicalTypeId, StrpTimeFormat> date_format = {{LogicalTypeId::DATE, {}}, {LogicalTypeId::TIMESTAMP, {}}};
```
Any specific reason to use map over unordered_map?

hawkfish replied:
Yeah, I couldn't see any `hash` definition for `LogicalTypeId` (which is really uint8_t`?). I can poke at that some more.( I first considered using just a `vector` with empty slots, but the `enum` is not dense.`)

Mytherin replied:
I think the map is fine, just wondering about the reason :) In general it should only contain two keys anyway.

Mytherin reviewed (commented):
Thanks for the PR! It looks very good. I would like to see more tests with CSVs containing each of the different date formats if possible. Also some tests with CSVs with conflicting date formats (e.g. containing only 01-01-2020). What happens in the case where there are multiple potential formats?

Otherwise I think it looks great. Thanks again :)

hawkfish commented:
I can add some more tests, no problem. Also, we can add `TIME` formats if you like as it is all table driven, but I don't think there's a lot to gain there.

The ambiguous format question is hard to test for because the output is the same either way. The code I wrote has prioritised formats, but the example you give would parse correctly for either `%m-%d-%Y` or` %d-%m-%Y`. Is there a way to dump the format used by `COPY` to the output?

Mytherin commented:
> I can add some more tests, no problem. Also, we can add `TIME` formats if you like as it is all table driven, but I don't think there's a lot to gain there.
> 
> The ambiguous format question is hard to test for because the output is the same either way. The code I wrote has prioritised formats, but the example you give would parse correctly for either `%m-%d-%Y` or` %d-%m-%Y`. Is there a way to dump the format used by `COPY` to the output?

Indeed 01-01-2020 would parse the same, but 01-02-2020 then :) I think rather than prioritised formats it might be better to throw an exception on an ambiguous date format being encountered, or falling back to parsing as a string so the user can later correct the parsing manually. 

There is no way to output the format used by the copy command at this point, although that might be useful for the auto detection indeed. Perhaps something along those lines could be added.

Commit:
Issue 816: Auto DATE tests

Add tests for all detected formats.
Move the %y formats ahead of the
%Y formats and change the ISO 8601
detection to match.

hawkfish committed:
Merge pull request #17 from cwida/master

More UBSAN fixes and Travis test run (#902)

Commit:
Merge branch 'master' into hawkfish-csv-dates

Mytherin commented on src/execution/operator/persistent/buffered_csv_reader.cpp:
```diff
@@ -455,6 +550,61 @@ vector<LogicalType> BufferedCSVReader::SniffCSV(vector<LogicalType> requested_ty
+							SetDateFormat(type_format_candidates.back(), sql_type.id());
+						}
+						// check all formats and keep the first one that works
+						int32_t date_parts[6];
```
This should be [7], not [6] :)

hawkfish replied:
Another victory for strong typing... We should change the argument to Parse to be a struct that has all the output variables. But that is a problem for another day.

Mytherin replied:
Indeed, good point.

Mytherin commented:
Thanks for the extra tests! I investigated the 32-bit Windows crash, and it seems to be an off-by-one in the date_parts array. As to why that bug only triggers on Windows 32-bit, I have no clue. Could you fix that? Then after the tests pass I think we can merge it.

hawkfish commented:
Another victory for strong typing... We should change the argument to `Parse` to be a `struct` that has all the output variables. But that is a problem for another day.

Commit:
Issue 816: Parse date_parts size

Fix calling sequence bug.
